client/src/components/blog/BlogForm.tsx

```diff
@@ -0,0 +1,278 @@
+    submitLabel?: string;
+    isEdit?: boolean;
+}
+
+const effortLevels: EffortLevel[] = ["Low", "Medium", "High"];
+const visibilities: Visibility[] = ["Public", "Private", "Friends"];
+
+const BlogForm = <T extends BlogCreate | BlogUpdate = BlogCreate | BlogUpdate>({ initialValues = {}, onSubmit, submitLabel = "Create" }: BlogFormProps<T>) => {
+    type CategoryType = string[];
+    interface FormState {
+        title: string;
+        total_cost?: number;
+        total_duration?: number;
+        effort_level?: EffortLevel;
+        place_id?: string;
+        categories: CategoryType;
+        visibility: Visibility;
+        description: string;
+    }
+    const [form, setForm] = useState<FormState>({
+        title: initialValues.title || "",
+        total_cost: initialValues.total_cost,
+        total_duration: initialValues.total_duration,
+        effort_level: initialValues.effort_level,
+        place_id: initialValues.place_id,
+        categories: (initialValues.categories as string[]) || [],
+        visibility: initialValues.visibility || "Public",
+        description: initialValues.description || "",
+    });
+    const [selectedPlace, setSelectedPlace] = useState<PlaceDto | null>(null);
+    const [categoryInput, setCategoryInput] = useState("");
+    const [error, setError] = useState<string | null>(null);
+
+    const handlePlaceSelect = (place: PlaceDto | null) => {
+        setSelectedPlace(place);
+        setForm((prev) => ({
+            ...prev,
+            place_id: place?.place_id || undefined,
+        }));
+    };
+
+    useEffect(() => {
+        const loadInitialPlace = async () => {
+            if (initialValues.place_id && !selectedPlace) {
+                try {
+                    const place = await getPlaceById(initialValues.place_id);
+                    setSelectedPlace(place);
+                } catch (error) {
+                    console.error('Error loading place:', error);
+                }
+            }
+        };
+        loadInitialPlace();
+    }, [initialValues.place_id, selectedPlace]);
+
+    const handleChange = (
+        e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement>
+    ) => {
+        const { name, value, type } = e.target;
+        setForm((prev) => ({
+            ...prev,
+            [name]: type === "number" && value !== "" ? Number(value) : value,
+        }));
+    };
+
+    const handleCategoryAdd = () => {
+        const cat = categoryInput.trim();
+        if (cat && !form.categories.includes(cat)) {
+            setForm((prev) => ({
+                ...prev,
+                categories: [...prev.categories, cat],
+            }));
+            setCategoryInput("");
+        }
+    };
+
+    const handleCategoryRemove = (cat: string) => {
+        setForm((prev) => ({
+            ...prev,
+            categories: prev.categories.filter((c) => c !== cat),
+        }));
+    };
+
+    const handleSubmit = (e: React.FormEvent) => {
```

Comment:
I think you need to work more on code optimization and clarity


```suggestion
    if (!form.title?.trim()) {
        setError("Title is required.");
        return;
    }
    
    setError(null);

    // Prepare submission data
    const submissionData = {
        ...form,
        place_id: form.place_id?.trim() || undefined,
        total_cost: form.total_cost != null ? Number(form.total_cost) : undefined,
        total_duration: form.total_duration != null ? Number(form.total_duration) : undefined,
    };

```